### ISSUE TITLE
Improve performance of high cardinality grouping by reusing hash values

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#11680]
### Is your feature request related to a problem or challenge?

As described on https://github.com/apache/datafusion/issues/11679, we can do better for high cardinality aggregates

One thing that consumes significant time in such queries is hashing, and I think we can reduce that significantly.

Specifically, for the multi-phase repartition plan, the number of hashed rows is something like 

```
(input cardinality)  + 2 * (intermediate group cardinality) * (number of partitions)
```

For low cardinality aggregates (e.g when the intermediate group cardinality is 1000) the second term is small (a few thousand extra hashes isn't a big deal)

However, for high cardinality aggregates (eg. when the intermediate cardinality is like 1,000,000 and there are 16 partitions) the second term is substantial

In pictures, this looks like

```
               ▲                          ▲
               │                          │
               │                          │
               │                          │
               │                          │
               │                          │
   ┌───────────────────────┐  ┌───────────────────────┐       4. The  AggregateMode::Final
   │GroupBy                │  │GroupBy                │       GroupBy computes hash(group keys)
   │(AggregateMode::Final) │  │(AggregateMode::Final) │       *AGAIN* to find the correct hash
   │                       │  │                       │       bucket
   └───────────────────────┘  └───────────────────────┘
               ▲                          ▲
               │                          │
               └─────────────┬────────────┘
                             │
                             │
                             │
                ┌─────────────────────────┐                   3. The output of the first phase
                │       Repartition       │                   is repartitioned by computing
                │         HASH(x)         │                   hash(group keys) -- this is the
                └─────────────────────────┘                   same hash as computed in step 2.
                             ▲
                             │
             ┌───────────────┴─────────────┐
             │                             │
             │                             │
┌─────────────────────────┐  ┌──────────────────────────┐     2. Each AggregateMode::Partial
│        GroubyBy         │  │         GroubyBy         │     GroupBy hashes the group keys to
│(AggregateMode::Partial) │  │ (AggregateMode::Partial) │     find the correct hash bucket.
└─────────────────────────┘  └──────────────────────────┘
             ▲                             ▲
             │                            ┌┘
             │                            │
        .─────────.                  .─────────.
     ,─'           '─.            ,─'           '─.
    ;      Input      :          ;      Input      :          1. Input is read
    :   Partition 0   ;          :   Partition 1   ;
     ╲               ╱            ╲               ╱
      '─.         ,─'              '─.         ,─'
         `───────'                    `───────'
```

This effect can be seen in profiling for click bench XXX:

(TODO screen shot showing large amounts of work done in hashing)

### Describe the solution you'd like

The basic idea is rather than rather than recompute the hash values in `RepartitionExec` and `AggregateMode::Final` we would reuse the values from `AggregateMode::Partial` (which has already computed a hash value for each input group)

Something like this
```text
                         ▲                          ▲                                                       
                         │                          │                                                       
                         │                          │                                                       
                         │                          │                                                       
                         │                          │                                                       
                         │                          │                                                       
             ┌───────────────────────┐  ┌───────────────────────┐       4. The  AggregateMode::Final        
             │GroupBy                │  │GroupBy                │       GroupBy also gets the hash values   
             │(AggregateMode::Final) │  │(AggregateMode::Final) │       and does not recompute them         
             │                       │  │                       │                                           
             └───────────────────────┘  └───────────────────────┘                                           
               ▲         ▲                          ▲                                                       
               │         │                          │                                                       
                         └─────────────┬────────────┘                                                       
Pass hash      │                       │                                                                    
values up the                          │                                                                    
plan rather    │                       │                                                                    
than                      ┌─────────────────────────┐                   3. In addition to the partial       
recomputing    │          │       Repartition       │                   aggregates and group values, *ALSO* 
them                      │    PRECOMPUTED_HASH     │                   pass the hash values to the         
               │          └─────────────────────────┘                   RepartitionExec which also passed   
                                       ▲                                them on to the AggregateMode::Final 
               │                       │                                                                    
                       ┌───────────────┴─────────────┐                                                      
               │       │                             │                                                      
                       │                             │                                                      
          ┌─────────────────────────┐  ┌──────────────────────────┐     2. Each AggregateMode::Partial      
          │        GroubyBy         │  │         GroubyBy         │     GroupBy hashes the group keys to    
          │(AggregateMode::Partial) │  │ (AggregateMode::Partial) │     find the correct hash bucket.       
          └─────────────────────────┘  └──────────────────────────┘                                         
                       ▲                             ▲                                                      
                       │                            ┌┘                                                      
                       │                            │                                                       
                  .─────────.                  .─────────.                                                  
               ,─'           '─.            ,─'           '─.                                               
              ;      Input      :          ;      Input      :          1. Input is read                    
              :   Partition 0   ;          :   Partition 1   ;                                              
               ╲               ╱            ╲               ╱                                               
                '─.         ,─'              '─.         ,─'                                                
                   `───────'                    `───────'                                                   
```

### Describe alternatives you've considered

We maybe could pass the data as an explicit new column somehow, or maybe as a field in a struct array 🤔

**Comment by alamb** [comment]
The trick with this ticket will be to structure the code in a way that is general and works across plans. 

It might first be worth a POC / hack to see how much performance there is to be had here (I suspect it is like 5-10% at most)

**Comment by jayzhan211** [comment]
The experiment I did in #11708 shows that
1. There is no much difference for clickbench Q17
2. Outperform for high cardinality, row num 2,000,000 with all the value is different
3. Simplify Repartition Hash code largely

@alamb If the benchmark code looks good to you, I think we could reuse hash
To further improve clickbench Q17, the bottleneck is now arrow::Row (RowConverter::append, and Rows::push), do you think there is room for improvement? or should we find a way to reduce Rows by design

**Comment by alamb** [comment]
Thank you @jayzhan211  -- that is some interesting results. 

I think it makes sense that reusing the hash values is helpful mostly for high cardinality aggregates as in that case the number of rows that need to be repartitioned /rehashed is high.

> Alternative idea for improvement is, if we can combine partial group + repartition + final group in one operation. We could probably avoid converting to row once again in final group.

I think this is the approach taken by systems like DuckDB as I understand it and I think it is quite intregruing to consider

The challenge of the approach would be the software engineering required to manage the complexity of the combined multi-stage operator. I am not sure the functioanlity would be easy to combine without some more refactoring 🤔

**Comment by alamb** [comment]
@Dandandan  has a good point on https://github.com/apache/datafusion/pull/11708#issuecomment-2262084412 that is some cases (like a network shuffle) passing the hash values might be more expensive than just recomputing them

**Comment by jayzhan211** [comment]
I found I got performance boost for clickbench Q17 just by enforcing single mode for group by, interesting

```rust
Comparing main and single-multi-groupby
--------------------
Benchmark clickbench_1.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃      main ┃ single-multi-groupby ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 0     │ 2009.51ms │            1435.19ms │ +1.40x faster │
│ QQuery 1     │ 7270.86ms │            4033.52ms │ +1.80x faster │
└──────────────┴───────────┴──────────────────────┴───────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━┓
┃ Benchmark Summary                   ┃           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━┩
│ Total Time (main)                   │ 9280.37ms │
│ Total Time (single-multi-groupby)   │ 5468.70ms │
│ Average Time (main)                 │ 4640.19ms │
│ Average Time (single-multi-groupby) │ 2734.35ms │
│ Queries Faster                      │         2 │
│ Queries Slower                      │         0 │
│ Queries with No Change              │         0 │
└─────────────────────────────────────┴───────────┘
```

https://github.com/apache/datafusion/pull/11762

I think what I need to do is find a query that is currently slower in single mode, and find a way to optimize it like partial/final way in single execution node? 🤔 

Does the result shows that what we really need is storing values in one large hash table + spilling?

**Comment by alamb** [comment]
> #11762
> 
> I think what I need to do is find a query that is currently slower in single mode, and find a way to optimize it like partial/final way in single execution node? 🤔

What exactly does `single_mode` do? https://github.com/apache/datafusion/pull/11762 looks like maybe it just uses a single group by node?


> Does anyone know what kind of query that is what partial/final group by good at?


I think they are good at being able to use multiple to do the work in parallel

They are especially good at low cardinality aggregates (some of the TPCH ones for example where there are 4 distinct groups) as the hash tables are small and the final shuffle is very small.

**Comment by jayzhan211** [comment]
> What exactly does single_mode do? https://github.com/apache/datafusion/pull/11762 looks like maybe it just uses a single group by node?

Yes, partial/final often do the hashing / row converter twice. In single group by node, there is only one.

**Comment by jayzhan211** [comment]
Difference between #11762 and main

#11762 runs Repartition -> SingleMode group by
`main` runs Partial group by-> Repartition -> Final group by

For high cardinality (2M rows unique values)
#11762 repartitioned to 12, divided equally, each has around ~166000 rows to do single group by (same with final step)
`main` run partial aggregate first, there are 256 partition with ~8000 rows each. 256 * 8000 is ~2M.

#11762 has less partition, and thus speed up a lot

For low cardinality (2M rows with only 4 values)
#11762 has 4 partitions, each 0.5M rows
`main` run partial aggregate and got 4 partition after repartition, only 1 row to process in final step

I can see #11762 is slightly slower than `main` for low cardinality case but the regression is negligible compare to high cardinality case.


Next, I want to move repartition within single mode group by. I guess we can see on par result for low cardinality case

**Comment by alamb** [comment]
I think this data is very interesting and we should look more deeply into why is the single group mode faster than doing a repartition / aggregate.

It seems like the only differences are:
1. There is a `RepartitionExec` and `CoalesceBatchesExec`
2. The final `AggregateExec` happens in parallel  (but on distinct subsets of the group)

I would expect doing the final aggregate in parallel on distinct subsets to be about as fast

So one reasonable conclusion conclusion that the overhead of `RepartitionExec` and `CoalesceBatchesExec` accounts for the difference 🤔 and this if we reduced the Repartition overhead we could see similar improvements as the group by single mode 


This is the idea behind exploring  https://github.com/apache/datafusion/pull/11647 -- I think we could avoid a copy at the output of CoalesceBatchesExec which would help to reduce the overhead

**Comment by Rachelint** [comment]
> I think this data is very interesting and we should look more deeply into why is the single group mode faster than doing a repartition / aggregate.
> 
> It seems like the only differences are:
> 
>     1. There is a `RepartitionExec` and `CoalesceBatchesExec`
> 
>     2. The final `AggregateExec` happens in parallel  (but on distinct subsets of the group)
> 
> 
> I would expect doing the final aggregate in parallel on distinct subsets to be about as fast
> 
> So one reasonable conclusion conclusion that the overhead of `RepartitionExec` and `CoalesceBatchesExec` accounts for the difference 🤔 and this if we reduced the Repartition overhead we could see similar improvements as the group by single mode
> 
> This is the idea behind exploring #11647 -- I think we could avoid a copy at the output of CoalesceBatchesExec which would help to reduce the overhead

It seems the cpu cost about `RepartitionExec` and `CoalesceBatchesExec` is not the bottleleck for the q32 in clickbench according to the flamegraph.

One possibility is that it may not be the problem CPU cost, and it is the problem about schedule (tokio).

![xxx](https://raw.githubusercontent.com/Rachelint/drawio-store/main/dperf.08085.svg)

**Comment by alamb** [comment]
> One possibility is that it may not be the problem about CPU cost, and it is the problem about schedule？

I am not sure -- is there any chance you can attach the `svg` version of the flamegraph so I can zoom in and look at it more carefully?

**Comment by Rachelint** [comment]
> > One possibility is that it may not be the problem about CPU cost, and it is the problem about schedule？
> 
> I am not sure -- is there any chance you can attach the `svg` version of the flamegraph so I can zoom in and look at it more carefully?

Ok, svg src here:
https://github.com/Rachelint/drawio-store/edit/main/dperf.08085.svg

**Comment by alamb** [comment]
> > > One possibility is that it may not be the problem about CPU cost, and it is the problem about schedule？
> > 
> > 
> > I am not sure -- is there any chance you can attach the `svg` version of the flamegraph so I can zoom in and look at it more carefully?
> 
> Ok, svg here: https://github.com/Rachelint/drawio-store/blob/738d864dd1fb0eb4639ddc1d1c18f87877c3e7a4/dperf.08085.svg

That is quite cool -- thank you @Rachelint 



I see evidence of why you proposed https://github.com/apache/arrow-rs/issues/6146 / https://github.com/apache/arrow-rs/pull/6155 🏃 
![Screenshot 2024-08-08 at 5 05 45 PM](https://github.com/user-attachments/assets/31721f67-8d17-4cc8-b655-f120d131ee81)
![Screenshot 2024-08-08 at 5 06 51 PM](https://github.com/user-attachments/assets/3a48cc2b-37ca-4fff-a7bf-126cfbdd3162)

**Comment by Rachelint** [comment]
@alamb Yes... the eager computation of null count is actually not as cheap as we expect...

**Comment by yjshen** [comment]
> It seems the cpu cost about RepartitionExec and CoalesceBatchesExec is not the bottleleck for the Q32

What @jayzhan211 experiments and shows the effects of single aggregate performance benefits in #11762 and #11777 is on Clickbench Q17/Q18 instead of Q32. 

As of today, I see that Q32 performance is comparable to that in DuckDB on an M3 Mac.

```
# DuckDB Q32:
0.5369797919993289
0.44854350000969134
0.41927954100538045

# DataFusion main(780cccb52)
0.620
0.400
0.409
```

But for Q17, we are still behind:

```
# DuckDB
0.5953990409907419
0.5309897500119405
0.5242392499931157

# DataFusion main(780cccb52)
1.145
1.072
1.082
```

We would probably need to consolidate Aggregate(Partial and Final) and Repartition into a single place in order to be able to adaptively choose aggregate mode/algorithm based on runtime statistics.

**Comment by Rachelint** [comment]
> > It seems the cpu cost about RepartitionExec and CoalesceBatchesExec is not the bottleleck for the Q32
> 
> What @jayzhan211 experiments and shows the effects of single aggregate performance benefits in #11762 and #11777 is on Clickbench Q17/Q18 instead of Q32.
> 
> As of today, I see that Q32 performance is comparable to that in DuckDB on an M3 Mac.
> 
> ```
> # DuckDB Q32:
> 0.5369797919993289
> 0.44854350000969134
> 0.41927954100538045
> 
> # DataFusion main(780cccb52)
> 0.620
> 0.400
> 0.409
> ```
> 
> But for Q17, we are still behind:
> 
> ```
> # DuckDB
> 0.5953990409907419
> 0.5309897500119405
> 0.5242392499931157
> 
> # DataFusion main(780cccb52)
> 1.145
> 1.072
> 1.082
> ```
> 
> We would probably need to consolidate Aggregate(Partial and Final) and Repartition into a single place in order to be able to adaptively choose aggregate mode/algorithm based on runtime statistics.

I see the improvement about q32 in later pr #11792, and I guess the reason why performance improved may be simlar as the partial skipping? Maybe q17/q18 are improved due to different reason with q32?

I agree maybe we should perform the similar mechanism about select the merging mode dynamicly like `duckdb`.

**Comment by yjshen** [comment]
> guess the reason why performance improved may be simlar as the partial skipping?

I would say partial skipping only contributes a small portion of improvement to Q32.

```
set datafusion.execution.skip_partial_aggregation_probe_rows_threshold=10000000000;
set datafusion.execution.skip_partial_aggregation_probe_ratio_threshold=1.0;
explain analyze SELECT "WatchID", "ClientIP", COUNT(*) AS c, SUM("IsRefresh"), AVG("ResolutionWidth") FROM hits WHERE "SearchPhrase" <> '' GROUP BY "WatchID", "ClientIP" ORDER BY c DESC LIMIT 10;
```

I get:

```
AggregateExec: mode=Partial, gby=[WatchID@0 as WatchID, ClientIP@1 as ClientIP], aggr=[count(*), sum(hits.IsRefresh), avg(hits.ResolutionWidth)], metrics=[output_rows=13172392, elapsed_compute=1.403636372s, skipped_aggregation_rows=0, skipped_aggregation_computation_time=16ns]
```

Remove the `explain analyze`, and I get the performance number:

```
0.730
0.535
0.530
```

Since we use `x2.15` time (with DataFusion 40.0.0) compared to DuckDB, partial skipping may not be essential for Q32.

**Comment by jayzhan211** [comment]
I try an alternative way other than merging partial + final. Continues on the single mode but find a way to optimize on low cardinality case. 

I found the reason that slow down tcph Q1 is that we repartition the batch so there are **much more** batches with a smaller size to compute than neccessary.  I try to remove the repartition + coalesce batch and found out the performance improves!

```
// main
Aggregate Partial. (each batches size has around 16k)
    Repartition
       Coalesce
            Aggregate Final
```

```
// single mode only
Projection (each batches size has around 16k)
    Repartition (re-hash, each batch sizes has only 10k)
       Coalesce
            Aggregate Single Partition (more `intern` is called)
```

```
// single mode without repartition. This is also only half a time than main branch, since there is only single aggregate exec
Projection (each batches size has around 16k)
     Aggregate Single Partition (equivalent `intern` is called)
```

https://github.com/apache/datafusion/pull/12340

The next thing is to find out **when** and **how** should I avoid `EnforceDistribution` and bypass `SanityCheckPlan` if needed. And ensure there is no other queries slow down.

**Comment by Rachelint** [comment]
> > I guess the reason why performance improved may be simlar as the partial skipping
> 
> Yes, it is why I experiment with single mode, force to avoid partial and repartition stage for all query, sadly, this doesn't work well for low cardinality case
> 
> > We would probably need to consolidate Aggregate(Partial and Final) and Repartition into a single place in order to be able to adaptively choose aggregate mode/algorithm based on runtime statistics.
> 
> I agree, similar to my idea before.
> 
> > Alternative idea for improvement is, if we can combine partial group + repartition + final group in one operation. We could probably avoid converting to row once again in final group.
> 
> However, the refactor is quite challenging

For aggr, It may be used to perform the parallel merging in final aggr  from partial aggr.
In my knowledge, duck seems use partitioned hashtable to perform the similar mechanism?

**Comment by yjshen** [comment]
> Does anyone know what the rationale of having repartition + coalesce, what kind of query benefits from it

The primary reason is scalability. Efficient aggregation requires multi-core CPUs to process data in parallel. To facilitate this and prevent contention from multiple threads altering a single hash table simultaneously (often managed with locks), a repartition phase is introduced. 

This repartition allows each thread to perform aggregation independently. Furthermore, pre-aggregation is employed to conduct preliminary calculations to streamline repartitioning, significantly reducing the volume of data that needs repartitioning(in cases either the cadinatlity is low or there are several hot keys).

> The next thing is to find out when and how should I avoid EnforceDistribution and bypass SanityCheckPlan if needed. And ensure there is no other queries slow down.

This when-and-how problem is difficult for query engines because it requires foreknowledge of the data characteristics on grouping keys. It's even harder for DataFusion since we have very limited table metadata that could help us with this decision. 

In an adaptive approach, the aggregate operator could internally start with one method and move to another without interacting with other components(the physical optimizer and other physical operators), making it more feasible.

**Comment by Rachelint** [comment]
> > Does anyone know what the rationale of having repartition + coalesce, what kind of query benefits from it
> 
> The primary reason is scalability. Efficient aggregation requires multi-core CPUs to process data in parallel. To facilitate this and prevent contention from multiple threads altering a single hash table simultaneously (often managed with locks), a repartition phase is introduced.
> 
> This repartition allows each thread to perform aggregation independently. Furthermore, pre-aggregation is employed to conduct preliminary calculations to streamline repartitioning, significantly reducing the volume of data that needs repartitioning(in cases either the cadinatlity is low or there are several hot keys).
> 
> > The next thing is to find out when and how should I avoid EnforceDistribution and bypass SanityCheckPlan if needed. And ensure there is no other queries slow down.
> 
> This when-and-how problem is difficult for query engines because it requires foreknowledge of the data characteristics on grouping keys. It's even harder for DataFusion since we have very limited table metadata that could help us with this decision.
> 
> In an adaptive approach, the aggregate operator could internally start with one method and move to another without interacting with other components(the physical optimizer and other physical operators), making it more feasible.

Maybe reducing the cost of repartition is an alternative? I think the reason why `partial + final` lead to  performance regression in some situation is possible that the cost of splitting batches is greater than the improvement of parallel merging?

**Comment by alamb** [comment]
> I am trying a poc about fusing the repartition and partial aggregation.

That is an interesting idea (perhaps have the partial group by operator produce partitioned output somehow (as it already knows the hash values of each group 🤔 )

**Comment by Rachelint** [comment]
> > > Does anyone know what the rationale of having repartition + coalesce, what kind of query benefits from it
> > 
> > 
> > The primary reason is scalability. Efficient aggregation requires multi-core CPUs to process data in parallel. To facilitate this and prevent contention from multiple threads altering a single hash table simultaneously (often managed with locks), a repartition phase is introduced.
> > This repartition allows each thread to perform aggregation independently. Furthermore, pre-aggregation is employed to conduct preliminary calculations to streamline repartitioning, significantly reducing the volume of data that needs repartitioning(in cases either the cadinatlity is low or there are several hot keys).
> > > The next thing is to find out when and how should I avoid EnforceDistribution and bypass SanityCheckPlan if needed. And ensure there is no other queries slow down.
> > 
> > 
> > This when-and-how problem is difficult for query engines because it requires foreknowledge of the data characteristics on grouping keys. It's even harder for DataFusion since we have very limited table metadata that could help us with this decision.
> > In an adaptive approach, the aggregate operator could internally start with one method and move to another without interacting with other components(the physical optimizer and other physical operators), making it more feasible.
> 
> Maybe reducing the cost of repartition is an alternative? I think the reason why `partial + final` lead to performance regression in some situation is possible that the cost of splitting batches is greater than the improvement of parallel merging? I am trying a poc about fusing the repartition and partial aggregation.

It is exciting that this idea seems promising! 
I tried some related queries in my local, and it seems faster.
- Main
```
Q16: SELECT "UserID", "SearchPhrase", COUNT(*) FROM hits GROUP BY "UserID", "SearchPhrase" ORDER BY COUNT(*) DESC LIMIT 10;
Query 16 iteration 0 took 5537.5 ms and returned 10 rows
Query 16 iteration 1 took 5308.1 ms and returned 10 rows
Query 16 iteration 2 took 5499.6 ms and returned 10 rows
Query 16 iteration 3 took 5561.5 ms and returned 10 rows
Query 16 iteration 4 took 5380.3 ms and returned 10 rows

Q17: SELECT "UserID", "SearchPhrase", COUNT(*) FROM hits GROUP BY "UserID", "SearchPhrase" LIMIT 10;
Query 17 iteration 0 took 4961.5 ms and returned 10 rows
Query 17 iteration 1 took 4926.8 ms and returned 10 rows
Query 17 iteration 2 took 5000.1 ms and returned 10 rows
Query 17 iteration 3 took 4952.6 ms and returned 10 rows
Query 17 iteration 4 took 4922.3 ms and returned 10 rows
```

- The poc #12526 
```
Q16: SELECT "UserID", "SearchPhrase", COUNT(*) FROM hits GROUP BY "UserID", "SearchPhrase" ORDER BY COUNT(*) DESC LIMIT 10;
Query 16 iteration 0 took 5252.9 ms and returned 10 rows
Query 16 iteration 1 took 5092.9 ms and returned 10 rows
Query 16 iteration 2 took 5102.9 ms and returned 10 rows
Query 16 iteration 3 took 5181.0 ms and returned 10 rows
Query 16 iteration 4 took 5169.2 ms and returned 10 rows

Q17: SELECT "UserID", "SearchPhrase", COUNT(*) FROM hits GROUP BY "UserID", "SearchPhrase" LIMIT 10;
Query 17 iteration 0 took 4660.8 ms and returned 10 rows
Query 17 iteration 1 took 4619.3 ms and returned 10 rows
Query 17 iteration 2 took 4588.3 ms and returned 10 rows
Query 17 iteration 3 took 4717.4 ms and returned 10 rows
Query 17 iteration 4 took 4633.0 ms and returned 10 rows
```

I will check it more carefully again, and if it actually works, I will submit a formal PR.

**Comment by Rachelint** [comment]
I am sure that the performance improve now.
But I think we should push this forward after:
- #11943 merged, because this pr made big code changes too, may be clever to  avoid conflicts with this
- Refactor the aggregation codes. I think it is getting unmaintainable... Some issues pointed out this too https://github.com/apache/datafusion/issues/12335

**Comment by alamb** [comment]
> But I think we should push this forward after:

I agree -- thank you for all this POC work @Rachelint  -- very cool.

I personally plan to
1.  help @jayzhan211  get https://github.com/apache/datafusion/pull/12269 over (and any cleanup / follow on work) -- 10-20% improvements
2. Get `StringView` on by default -- another 10-20% improvements

Then I will have bandwidth to consider the intermediate blocked management

> Refactor the aggregation codes. I think it is getting unmaintainable... Some issues pointed out this too Improve aggregation code readability  #12335

Yes, I think this is the biggest challenge at the moment. I actually view the fuzz tesing as a critical piece of this so as we rearrange the code we have confidence we aren't introducing regressions

Thank you again for all your help so far

**Comment by Rachelint** [comment]
> > But I think we should push this forward after:
> 
> I agree -- thank you for all this POC work @Rachelint -- very cool.
> 
> I personally plan to
> 
>     1. help @jayzhan211  get [Avoid RowConverter for multi column grouping  #12269](https://github.com/apache/datafusion/pull/12269) over (and any cleanup / follow on work) -- 10-20% improvements
> 
>     2. Get `StringView` on by default -- another 10-20% improvements
> 
> 
> Then I will have bandwidth to consider the intermediate blocked management
> 
> > Refactor the aggregation codes. I think it is getting unmaintainable... Some issues pointed out this too Improve aggregation code readability  #12335
> 
> Yes, I think this is the biggest challenge at the moment. I actually view the fuzz tesing as a critical piece of this so as we rearrange the code we have confidence we aren't introducing regressions
> 
> Thank you again for all your help so far

Yes, I agree testing is extremely important before introducing further big changes to aggregation codes.

I am also researching and thinking how to refine the exists fuzz tests.

**Comment by jayzhan211** [comment]
> I am sure that the performance improved now.
> 
> But I think we should push this forward after:
> 
> * [Sketch for aggregation intermediate results blocked management #11943](https://github.com/apache/datafusion/pull/11943) merged, because this pr made big code changes too, may be clever to  avoid conflicts with this
> * Refactor the aggregation codes. I think it is getting unmaintainable... Some issues pointed out this too [Improve aggregation code readability  #12335](https://github.com/apache/datafusion/issues/12335)
>   I have some thoughts about the refactor [Improve aggregation code readability  #12335 (comment)](https://github.com/apache/datafusion/issues/12335#issuecomment-2358310830)

@Rachelint IMO, I think we could work on fusing Partial + Repartition first, I'm quite confident on the direction of this improvement.  To be honest I don't quite follow #11943 and it seems that we need better test coverage before #11943 be merged. I think we could work on other optimization first (fusing + simplify code #12335) + fuzz test improvement and then review #11943 again after that.

**Comment by Rachelint** [comment]
> > I am sure that the performance improved now.
> > But I think we should push this forward after:
> > 
> > * [Sketch for aggregation intermediate results blocked management #11943](https://github.com/apache/datafusion/pull/11943) merged, because this pr made big code changes too, may be clever to  avoid conflicts with this
> > * Refactor the aggregation codes. I think it is getting unmaintainable... Some issues pointed out this too [Improve aggregation code readability  #12335](https://github.com/apache/datafusion/issues/12335)
> >   I have some thoughts about the refactor [Improve aggregation code readability  #12335 (comment)](https://github.com/apache/datafusion/issues/12335#issuecomment-2358310830)
> 
> @Rachelint I suggest we work on fusing Partial + Repartition first, I'm quite confident on the direction of this improvement. To be honest I don't quite follow #11943 and it seems that we need better test coverage before #11943 be merged. I think we could work on other optimization first (fusing + simplify code #12335) + fuzz test improvement and then review #11943 again after that.

Thanks for suggestion! 
Seems it is actually clever for working `fusing Partial + Repartition` first.  Although the #11943 can improve more performance(10%~25%, and fusing `fusing Partial + Repartition` improve 5%~10%) but it indeed needs much larger code changes.

But after considering again, I guess it possible that fuzz test and code refactor may should be finished first before continuing to introduce more non-trivial changes? 

The aggregation codes are actually too complex now, and to be honest I also have no enough confidence to make big changes currently (actually I found possible bugs and regressions in main, and I am checking them again more carefully).

I plan to work on them when I have bandwidth after finishing the string related optimizations.

**Comment by alamb** [comment]
I think you should also not discount what you find personally exciting :) I realize that testing is not always to everyones liking

One thing to contemplate that I have found works well is to
1.  get the foundation setup (e.g. make the aggregation fuzz runner)
2. write a few cases yourself
3. enumerate the other cases 
4. Turn the community loose on implementing the other cases

I have found people seem very excited to help out, contribute to the project, and learn Rust and they do so when there is a clear description of what is needed and an example to follow.

**Comment by alamb** [comment]
I agree that fusing Partial + Repartition sounds like a nice win with lower code changes needed (and is covered well by existing tests)

I just merged https://github.com/apache/datafusion/pull/12269 and I will file some small follow ons related to that tomorrow. 

So exciting. It is great to be on the team with you @Rachelint and @jayzhan211 ❤️

**Comment by alamb** [comment]
@robtandy @adriangb @gabotechs @NGA-TRAN  this is the ticket I was referring to that describes reusing hash values between repartition and group by

It is probably worth pointing out here that this basic optimization applies to repartitioned join keys as well

**Comment by NGA-TRAN** [comment]
> It is probably worth pointing out here that this basic optimization applies to repartitioned join keys as well

@alamb : yes. This will help to improve https://github.com/datafusion-contrib/datafusion-distributed/issues/118 and https://github.com/datafusion-contrib/datafusion-distributed/issues/119

**Comment by adriangb** [comment]
The approach discussed today was that the GroupBy or HashJoin operator would inject an additional projection with the expression to compute the partition. This would then be computed all the way down into the scan potentially and be available for both the shuffle step and the operator itself without recomputing. The operator needs to then internally strip that additional column off before passing to its parent.

**Comment by Dandandan** [comment]
While reading a bit more on this topic, I am quite convinced that the approach (or a similar one like) https://github.com/apache/datafusion/issues/20773 is the way forward, instead of only re-using the hash value or optimizing the materialization after the partial aggregation.

**Comment by gabotechs** [comment]
I'm giving this a try in https://github.com/apache/datafusion/pull/23720, and even if there are no life-changing improvements, it's looking good so far.

Still iterating over it, but if everything goes well it should be ready for review soon.

**Comment by Rachelint** [comment]
I also tried this in #23446 before.
And for clickbench, it seems that only can get improvement in cases with expansive hash calculation.
I think the reason is that:
- Creating and operating the hash column(a Uint64Array) may be not trivial as we think
- And also hash calculation for `primitive`s may be also not expansive as we think

About how to define `cases with expansive hash calculation`, I think maybe we can start from the static and simple rules:
- reuse when `byte/byte view` or `other more complex columns` appear in `group by`
- reuse when we group by a large amount of columns even all of them are `primitive`
- not reuse in other cases

**Comment by gabotechs** [comment]
🤔 Interesting, I'd have expected that maintaining an extra `Uint64Array` column would be pretty much 0 cost, as creating it costs compute that we anyways needed to do one way or another.

For the clickbench results in both PRs (https://github.com/apache/datafusion/pull/23720#issuecomment-5021747597 and https://github.com/apache/datafusion/pull/23446#issuecomment-4950574858), I see the gains and regressions are so subtle and inconsistent that I could believe everything there is within noise.

ClickBench q4 seems to be consistent in regressing, I'll pass that through a profiler and see if there's something obvious going on.

**Comment by gabotechs** [comment]
I've been trying it some more, and ended up closing the PR as it did not prove any meaningful performance improvements, not even being smart and just apply hash reusing for non-primitive columns.

Left some conclusion here:
- https://github.com/apache/datafusion/pull/23720

At least with the public benchmarks, hashing occupies a very small fraction of the profiles, and even avoiding half the hashing operations in a query only shows some improvements that could perfectly fall within noise.

My take is that this is a high complexity / low impact piece of work that could probably be worth exploring once other clearer performance optimizations have been addressed.
